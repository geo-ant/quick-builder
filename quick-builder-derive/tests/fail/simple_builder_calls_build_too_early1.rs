use quick_builder_derive::QuickBuilder;

#[derive(QuickBuilder, PartialEq, Debug)]
#[invariant(|this|*this.first > this.third)]
struct Foo<'a, T> {
    #[invariant(|f|!f.is_nan())]
    first: &'a f32,
    second: T,
    third: f32,
}

fn foo() {
    let float = 1.;

    // this should fail because the builder method is only available after
    // third(...) has been called.
    let built = Foo::<i32>::builder()
        .first(&float)
        // .second(10)
        // .third(0.3)
        .build();
    // .expect("building this must not fail");
}

pub fn main() {}
