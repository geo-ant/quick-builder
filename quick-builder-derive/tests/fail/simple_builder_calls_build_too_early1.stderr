error[E0599]: no method named `build` found for struct `FooBuilder<'_, i32, (&f32,)>` in the current scope
  --> tests/fail/simple_builder_calls_build_too_early1.rs:21:10
   |
3  |   #[derive(QuickBuilder, PartialEq, Debug)]
   |            ------------ method `build` not found for this struct
...
17 |       let built = Foo::<i32>::builder()
   |  _________________-
18 | |         .first(&float)
19 | |         // .second(10)
20 | |         // .third(0.3)
21 | |         .build();
   | |         -^^^^^ method not found in `FooBuilder<'_, i32, (&f32,)>`
   | |_________|
   |
   |
   = note: the method was found for
           - `FooBuilder<'a, T, (&'a f32, T, f32)>`
